### PR TITLE
feat(indexer): realtime updates of fills and bars

### PR DESCRIPTION
**Pull request by lidatong**
publishes fills and realtime bars into Redis

Screen Shot 2022-10-24 at 10 35 33 PM
Screen Shot 2022-10-24 at 10 35 50 PM

### CODE REVIEW

**Review by yuhanfang — APPROVED** [review]

File: aptos/api/aux-ts/src/indexer/index.ts [hunk 1]
```diff
@@ -86,188 +87,161 @@ async function publishAmmEvents() {
...
+    marketInput.baseCoinType !== auxClient.getWrappedFakeCoinType(FakeCoin.ETH)
+  ) {
+    return;
+  }
+  if (
+    marketInput.quoteCoinType !==
+    auxClient.getWrappedFakeCoinType(FakeCoin.USDC)
+  ) {
+    return;
+  }
```
Comment (leina05) [inline aptos/api/aux-ts/src/indexer/index.ts:130]: Still need these checks?

File: aptos/api/aux-ts/src/indexer/index.ts [hunk 2]
```diff
@@ -86,188 +87,161 @@ async function publishAmmEvents() {
...
+  const time = now - delta;
+
+  const len = await redisClient.lLen(key);
+  const rawFills = await redisClient.lPopCount(key, len) ?? [];
+  const fills: Order[] = rawFills
+    .map((s) => JSON.parse(s))
+    .filter((fill: Order) => Number(fill.time) >= time);
 
+  if (_.isEmpty(fills)) {
+    await redisPubSub.publish("BAR", { resolution, time, ohlcv: null });
```
Comment (leina05) [inline aptos/api/aux-ts/src/indexer/index.ts:220]: Why do we need to publish here (as opposed to no op)?

File: aptos/api/aux-ts/src/indexer/index.ts [hunk 3]
```diff
@@ -86,188 +87,161 @@ async function publishAmmEvents() {
...
+      close: _.last(prices)!,
+      volume: _(fills)
+        .map((fill) => fill.price * fill.quantity)
+        .sum(),
+    },
   };
-
-  fs.writeFileSync(PATH, JSON.stringify(bar));
-  return bar;
+  await redisPubSub.publish("BAR", bar);
```
Comment (leina05) [inline aptos/api/aux-ts/src/indexer/index.ts:237]: It seems like you're publishing bars for every resolution once per second. Is trading view able to resolve that into bars of various resolutions? E.g., for 1m resolution does it take the last bar update from every minute as the historical bar?

**Review by leina05 — APPROVED** [review]
Looks good! Maybe updating the indexer README would be helpful as a follow-up. Would be nice to eventually have fills as time series data in clickhouse to avoid some of the redis maneuvering

File: aptos/api/aux-ts/src/indexer/index.ts (hunk 2 above)
Comment (lidatong) [reply aptos/api/aux-ts/src/indexer/index.ts:220]: good q, i was considering the same and it's just an explicit null for that time vs. an implicit time gap

File: aptos/api/aux-ts/src/indexer/index.ts (hunk 3 above)
Comment (lidatong) [reply aptos/api/aux-ts/src/indexer/index.ts:275]: yep, every bar is published with a "time" which is the start time of the interval for trading view to use

we might publish the same bar multiple times (e.g. every second for those 15 seconds) as an updated view

moreover, there's of course a time-gap between when it actually happened and when the producer polls it, so there's kinda the two axis of time processing vs. event time

File: aptos/api/aux-ts/src/graphql/typeDefs/market.graphql [hunk 4]
```diff
@@ -100,6 +100,16 @@ type Order {
   time: Timestamp!
 }
 
+type Trade {
```
[the remark this answers is not in the document]
Comment (leina05) [reply aptos/api/aux-ts/src/graphql/typeDefs/market.graphql:103]: Actually maybe not the same thing here.

File: aptos/api/aux-ts/src/indexer/index.ts (hunk 3 above)
Comment (leina05) [reply aptos/api/aux-ts/src/indexer/index.ts:269]: Oh I see, `time` is the start time. Makes sense.

**Review by leina05 — COMMENTED** [review]
LGTM

**Pull request merged**